### ISSUE TITLE
noexcept regression in 3.0.11

### ISSUE THREAD

**Issue by scoder** [cython/cython#6335]
The change in https://github.com/cython/cython/pull/6124 breaks a function that is implicit noexcept in the pxd file as in:
```
$ cat a.pxd 
cdef int f()
$ cat a.pyx 
cdef int f():
    pass
$ cython -3X legacy_implicit_noexcept=True a.pyx 
warning: a.pxd:1:10: Implicit noexcept declaration is deprecated. Function declaration should contain 'noexcept' keyword.
warning: a.pyx:1:10: env.entries[self.declared_name()]=Entry(<7f78f7ec0ce0>, name=f, type=int (void) noexcept). 

Error compiling Cython file:
------------------------------------------------------------
...
cdef int f():
     ^
------------------------------------------------------------

a.pyx:1:5: Function signature does not match previous declaration
```

In particular, sagemath fails to build with cython 3.0.11.

_Originally posted by @tornaria in https://github.com/cython/cython/issues/6124#issuecomment-2282329891_

**Comment by matusvalo** [comment]
The change in #6124 is not correct. Should be implemented in different way - see https://github.com/cython/cython/pull/6124#issuecomment-2282867768

**Comment by tornaria** [comment]
Here's my suggestion from https://github.com/cython/cython/pull/6124#issuecomment-2282955499.

First chunk is reverting #6124, second chunk is the proposed fix for #6122.

```diff
--- a/Cython/Compiler/Nodes.py
+++ b/Cython/Compiler/Nodes.py
@@ -710,10 +710,8 @@ class CFuncDeclaratorNode(CDeclaratorNode):
                 and not self.has_explicit_exc_clause
                 and self.exception_check
                 and visibility != 'extern'):
-            # If function is already declared from pxd, the exception_check has already correct value.
-            if not (self.declared_name() in env.entries and not in_pxd):
-                self.exception_check = False
             # implicit noexcept, with a warning
+            self.exception_check = False
             warning(self.pos,
                     "Implicit noexcept declaration is deprecated."
                     " Function declaration should contain 'noexcept' keyword.",
@@ -3128,6 +3126,7 @@ class DefNode(FuncDefNode):
             if scope is None:
                 scope = cfunc.scope
             cfunc_type = cfunc.type
+            has_explicit_exc_clause=True
             if len(self.args) != len(cfunc_type.args) or cfunc_type.has_varargs:
                 error(self.pos, "wrong number of arguments")
                 error(cfunc.pos, "previous declaration here")
```
